inc/zoo/swar/SWAR.h

```diff
@@ -85,15 +85,24 @@ struct SWAR {
         LowerBits = MostSignificantBit - LeastSignificantBit,
         MaxUnsignedLaneValue = LeastSignificantLaneMask;
 
-    template <typename U>
-    constexpr static auto from_array(const U (&values)[Lanes]) noexcept {
+    template <typename InputIt>
+    constexpr static auto from_range(InputIt first, InputIt last) noexcept {
         auto result = T{0};
-        for (auto value : values) {
-            result = (result << NBits) | value;
+        for (; first != last; ++first) {
+            result = (result << NBits) | *first;
         }
-        return SWAR{result};
+        return result;
     }
 
+    template <typename U>
+    constexpr static auto from_array(const U (&values)[Lanes]) noexcept {
+        return SWAR{from_range(std::begin(values), std::end(values))};
+    }
+
+    using ArrayType = std::array<T, Lanes>;
+
+    constexpr SWAR(const ArrayType& array) : m_v{from_range(array.begin(), array.end())} {}
```

Comment:
You're again using L-value references gratuitously.
Let's say you're passed a temporary `std::vector` , since you explicitly request the L-value reference here, a totally unnecessary copy of the vector will be made, and the compiler won't save you from the folly since an allocation is potentially involved, the compiler won't be able to prove no allocation will be called and then it will have to preserve the potential side effects of the allocation.

Reply:
Please put the identifier modifiers next to the identifier they modify.
`ArrayType &array`
I don't know why, given the rules of the language, the style of putting the modifiers to the type is overwhelmingly more popular, since they don't bind to the type, but to the identifier being declared, they simply don't apply to the type!